Feature/support reading pem files from dir

mikonieminen opened the pull request:
This change is build on top of https://github.com/oxidizing/letters/pull/14 that needs to be fixed first and then this one should be rebased on top of that one.

mikonieminen committed:
Add support for reading CA certs from folder

Provide support for reading multiple PEM files from a folder when
reading in CA certificates for authenticating peer connection (server).

This also names the two functions in the Config module to match with
naming used by cURL.

joseferben commented on lib/letters.ml:
```diff
@@ -72,14 +88,28 @@ let load_file path =
+  |> Lwt_stream.to_list
+
+let certs_of path =
+  let ( <.> ) f g x = f (g x) in
```
Imo this adds more cognitive load than it reduces

mikonieminen replied:
Fixed. Now this code is gone. I managed to find significantly simpler approach.

joseferben commented on lib/letters.ml:
```diff
@@ -72,14 +88,28 @@ let load_file path =
+  load_file path
+  (* TODO add support for .der files *)
+  >|= (X509.Certificate.decode_pem_multiple <.> Cstruct.of_string)
+  >|= Rresult.R.get_ok
```
Why use an implicit dependency? Why not just use `Result.ok`?

mikonieminen replied:
Fixed. Now this code is gone. I managed to find significantly simpler approach.

joseferben commented on lib/letters.ml:
```diff
@@ -72,14 +88,28 @@ let load_file path =
+  >|= Rresult.R.get_ok
+
 let certs_of_pem path =
   let ( <.> ) f g x = f (g x) in
```
Same issue as above but additionally it is defined here twice.

mikonieminen replied:
Fixed. Now this code is gone. I managed to find significantly simpler approach.

joseferben commented on lib/letters.ml:
```diff
@@ -140,30 +170,66 @@ let build_email ~from ~recipients ~subject ~body =
       let multipart_content_alternative =
         let open Content_type in
-        Content_type.make `Multipart (Subtype.v `Multipart "alternative") Parameters.empty
+        Content_type.make `Multipart
```
`Content_type` is opened but then still referenced

mikonieminen replied:
This is not really related to this PR, but fixed anyway.

joseferben commented on lib/letters.ml:
```diff
@@ -140,30 +170,66 @@ let build_email ~from ~recipients ~subject ~body =
 
-let send ~config:c ~recipients:r ~message:m =
+let ca_cert_peer_verifier path =
+  let ( let* ) = Lwt.bind in
```
No consistent use of Lwt.bind, in code above the infix operators were used. I suggest to use `let ( let* ) = Lwt.bind` everywhere, it makes the code less noisy. At the top `open Lwt.Syntax` can be put once since it is a really common thing to use. Please replace infix bind with it as well.

mikonieminen replied:
Fixed.

joseferben replied:
Can you just do `open Lwt.Syntax` at the top? It seems like that is the way to go in other more recent libraries.

mikonieminen replied:
I'll think about it. I'm not sure if I like it, but on the other hand, I see the point. But not going to change it in this PR.

joseferben commented on lib/letters.ml:
```diff
@@ -140,30 +170,66 @@ let build_email ~from ~recipients ~subject ~body =
+  let* certs =
+    match Fpath.of_string path with
+    | Ok path -> certs_of path
+    | Error _ -> failwith "Failed to read CA certificate file"
```
Don't throw a string, but make a custom exception that you throw.

mikonieminen replied:
Fixed, I found a way to simplify the code quite significantly.

joseferben commented on lib/letters.ml:
```diff
@@ -140,30 +170,66 @@ let build_email ~from ~recipients ~subject ~body =
+  let* certs =
+    match Fpath.of_string path with
+    | Ok path -> certs_of_pem_directory path
+    | Error _ -> failwith "Failed to open CA certificates directory"
```
Don't throw a string, but make a custom exception that you throw.

mikonieminen replied:
Fixed, I found a way to simplify the code quite significantly.

joseferben reviewed (changes requested):
The error handling seems a bit inconsistent, sometimes we throw and sometimes we return `Result.t`.

mikonieminen committed:
Fix unnecessary reference through module

One of the uses of `Content_type.make` it is referenced through the
module unnecessarily while the module is already open in the scope.

joseferben commented on lib/letters.ml:
```diff
@@ -205,16 +227,20 @@ let send ~config:c ~recipients:r ~message:m =
+        let* cert = Ca_certs.detect () in
+        match cert with
+        | Some (`Ca_file path) -> ca_cert_peer_verifier path
+        | None -> failwith "Could not find CA certificate bundle" )
```
Don't throw a string but a custom exception

mikonieminen replied:
Not really part of this PR, but I have created a new issue for this one.

joseferben replied:
Thanks!

joseferben commented on lib/letters.ml:
```diff
@@ -1,14 +1,38 @@
-}
-
-type body = Plain of string | Html of string | Mixed of string * string * (string option)
+module Config = struct
```
Please provide at least `equal` and `show`, ideally `pp` as well for easier debugging and testing.

mikonieminen replied:
This is actually part of the other PR, not this one. I'll create separate ticket to get forward with this.

joseferben replied:
Thanks!

joseferben commented:
Thanks!

aronerben commented on lib/letters.ml:
```diff
@@ -1,14 +1,38 @@
-
-type body = Plain of string | Html of string | Mixed of string * string * (string option)
+module Config = struct
+  type ca_certs = CA_bundle of string | CA_path of string | Detect
```
Please name these `Ca_bundle` and `Ca_path`

mikonieminen replied:
Nope, CA is a short and it is common practise not to drop the case like you asked. Also this is inline with X509 library's conventions.

joseferben replied:
It is not idiomatic if you look at other libs: https://ocsigen.org/tyxml/4.4.0/manual/intro
`Html_sig`, `Html_types`, `Svg_sig`, `Svg_types`. 
I have not seen this kind of naming in the wild in OCaml. I have seen it in Java world (XMLWriter) which arguably defeats the purpose of CamelCasing.

aronerben replied:
Yes, that's why I wrote this. We have stuff like `Ca_file` in the code that is coming from mirage.

mikonieminen replied:
Very common practise is to keep all abbreviations in all caps when you have 1-3 char abbreviation. Also, in `ocaml-x509`, they follow this practise (not sure what would thay do if they had 4 char abbreviation or longer, but it does not matter here). See: [x509.mli](https://github.com/mirleft/ocaml-x509/blob/master/lib/x509.mli)

It can be argued if this defeats CamelCasing since abbreviations are a bit special case. Keeping abbreviations in all uppercase up to 3 chars is not just Java convention, it is a bit older one and this topic has been argued in different forums as long as CamelCase as a concept has existed. Claiming that idiomatic OCaml is to follow one or another is a bit wrong claim since the both are used. Because having both letters in uppercase is what X509 library does, I would keep both in uppercase here too since this part is related to the use of X509 lib.

mikonieminen replied:
See more about this topic in [stackoverflow](https://stackoverflow.com/questions/15526107/acronyms-in-camelcase)

mikonieminen replied:
And by the way, this is not even CamelCase that OCaml is using here.

joseferben replied:
My example was Java specific to highlight this never ending discussion. Just saying that given the OCaml I have been reading so far, `CA_file` looks odd in my eyes. I suggest to follow conventions of Jane Street packages.

I have no strong feelings about which one to choose as long as we are consistent across our packages and as long as the rule we follow is simple and clear.

joseferben replied:
The x509.mli seems to be a bad example of consistent naming.
For variants they have stuff like `EDI_party`, `DNS`, `Rfc_822`, `Ipsec_end`

joseferben replied:
I see no value in using `CA_file` here and something `Html_parser` in some other package. So what is your rule? 2-3 letter abbreviations are all capital? This is less simple than just saying "first letter is capital", doesn't matter if abbreviation or not.

hannesm replied:
(co-author of X509 here) yes, I agree that its naming is inconsistent in X509. unfortunately renaming has a cost (including bumping the major version and fixing reverse dependencies). I'd go with `Ca_file`, `Ca_bundle`, etc. here. (guidance for choosing identifiers in OCaml: https://ocaml.org/learn/tutorials/guidelines.html#How-to-choose-identifiers)

joseferben replied:
@hannsem Thanks a lot for your work on X509 and your input! I guess the fact that the longest discussion in this review is about naming speaks for itself :smile:

mikonieminen replied:
It seems I'm overruled 1 to 3 and I need to submit to the majority opinion :smile: 

Fixing.

mikonieminen replied:
Fixed. I also change use of `bundle` to `cert` so that it is more consistent with rest of the code. So `CA_path` is now `Ca_path` and `CA_bundle` is `Ca_cert`.

aronerben replied:
Thanks!

aronerben reviewed (commented):
The rest looks good

mikonieminen committed:
Add missing API documentation for sender
